File: src/client/java/netxms-client/src/main/java/org/netxms/client/objects/Rack.java
/**
 * NetXMS - open source network management system
 * Copyright (C) 2003-2022 Victor Kirhenshtein
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package org.netxms.client.objects;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import org.netxms.base.NXCPCodes;
import org.netxms.base.NXCPMessage;
import org.netxms.client.NXCSession;
import org.netxms.client.objects.configs.PassiveRackElement;
import org.netxms.client.objects.interfaces.Assets;
import org.netxms.client.objects.interfaces.HardwareEntity;

/**
 * Rack object
 */
public class Rack extends GenericObject implements Assets
{
	private int height;
	private boolean topBottomNumbering;
   private List<PassiveRackElement> passiveElements;
   private Map<String, String> assetList;

	/**
    * Create from NXCP message.
    *
    * @param msg NXCP message
    * @param session owning client session
	 */
	public Rack(NXCPMessage msg, NXCSession session)
	{
		super(msg, session);
		height = msg.getFieldAsInt32(NXCPCodes.VID_HEIGHT);
		topBottomNumbering = msg.getFieldAsBoolean(NXCPCodes.VID_TOP_BOTTOM);
		passiveElements = new ArrayList<PassiveRackElement>();
		int count = msg.getFieldAsInt32(NXCPCodes.VID_NUM_ELEMENTS);
		long fieldId = NXCPCodes.VID_ELEMENT_LIST_BASE;
		for (int i = 0; i < count; i++)
		{
		   passiveElements.add(new PassiveRackElement(msg, fieldId));
         fieldId += 10;
		}
		
		assetList = msg.getStringMapFromFields(NXCPCodes.VID_AM_DATA_BASE, NXCPCodes.VID_AM_COUNT);
	}

	/**
	 * @see org.netxms.client.objects.AbstractObject#getObjectClassName()
	 */
	@Override
	public String getObjectClassName()
	{
		return "Rack";
	}

	/**
	 * @see org.netxms.client.objects.AbstractObject#isAllowedOnMap()
	 */
	@Override
	public boolean isAllowedOnMap()
	{
		return true;
	}

   /**
    * @see org.netxms.client.objects.AbstractObject#isAlarmsVisible()
    */
   @Override
   public boolean isAlarmsVisible()
   {
      return true;
   }

	/**
	 * @return the height
	 */
	public int getHeight()
	{
		return height;
	}
	
	/**
    * @return the topBottomNumbering
    */
   public boolean isTopBottomNumbering()
   {
      return topBottomNumbering;
   }
   
   /**
    * Get rack attribute config entry list
    * 
    * @return entryList
    */
   public List<PassiveRackElement> getPassiveElements()
   {
      return passiveElements;
   }

   /**
	 * Get rack units, ordered by unit numbers
	 * 
	 * @return rack units, ordered by unit numbers
	 */
	public List<HardwareEntity> getUnits()
	{
	   List<HardwareEntity> units = new ArrayList<HardwareEntity>();
	   for(AbstractObject o : getChildrenAsArray())
	   {
	      if (o instanceof HardwareEntity)
	         units.add((HardwareEntity)o);
	   }
	   Collections.sort(units, new Comparator<HardwareEntity>() {
         @Override
         public int compare(HardwareEntity e1, HardwareEntity e2)
         {
            return e1.getRackPosition() - e2.getRackPosition();
         }
      });
	   return units;
	}

	/**
    * Get passive element by ID.
    * 
    * @param id element ID
    * @return passive element with given ID or null if such element does not exist
    */
	public PassiveRackElement getPassiveElement(long id)
	{
	   for(PassiveRackElement e : passiveElements)
	      if (e.getId() == id)
            return e;
      return null;
	}

   /**
    * @see org.netxms.client.objects.interfaces.Assets#getAssets()
    */
   @Override
   public Map<String, String> getAssets()
   {
      return assetList;
   }
}

File: src/server/drivers/lenovo/rack.cpp
/*
** NetXMS - Network Management System
** Driver for Lenovo ThinkSystem rack switches
** Copyright (C) 2003-2026 Victor Kirhenshtein
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU Lesser General Public License as published by
** the Free Software Foundation; either version 3 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU Lesser General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
**
** File: rack.cpp
**
**/

#include "lenovo.h"

/**
 * Get driver name
 */
const TCHAR *LenovoRackDriver::getName()
{
   return _T("LENOVO-RACK");
}

/**
 * Get driver version
 */
const TCHAR *LenovoRackDriver::getVersion()
{
   return NETXMS_VERSION_STRING;
}

/**
 * Check if given device can be potentially supported by driver
 *
 * @param oid Device OID
 */
int LenovoRackDriver::isPotentialDevice(const SNMP_ObjectId& oid)
{
   return oid.startsWith({ 1, 3, 6, 1, 4, 1, 19046, 1, 7 }) ? 127 : 0;
}

/**
 * Check if given device is supported by driver
 *
 * @param snmp SNMP transport
 * @param oid Device OID
 */
bool LenovoRackDriver::isDeviceSupported(SNMP_Transport *snmp, const SNMP_ObjectId& oid)
{
   return true;
}

/**
 * Get list of interfaces for given node
 *
 * @param snmp SNMP transport
 * @param node Node
 * @param driverData driver-specific data
 * @param useIfXTable if true, usage of ifXTable is allowed
 */
InterfaceList *LenovoRackDriver::getInterfaces(SNMP_Transport *snmp, NObject *node, DriverData *driverData, bool useIfXTable)
{
   InterfaceList *ifList = NetworkDeviceDriver::getInterfaces(snmp, node, driverData, true);
   if (ifList == nullptr)
      return nullptr;

   uint32_t port = 1;
   for (int i = 0; i < ifList->size(); i++)
   {
      InterfaceInfo *iface = ifList->get(i);
      if (iface->type == IFTYPE_ETHERNET_CSMACD)
      {
         iface->isPhysicalPort = true;
         iface->location.chassis = 1;
         iface->location.module = 0;
         iface->location.port = port++;
      }
   }

   return ifList;
}
